pkg/pillar/hypervisor/kvm.go

```diff
@@ -515,15 +516,30 @@ func (ctx kvmContext) Start(domainName string, domainID int) error {
 }
 
 func (ctx kvmContext) Stop(domainName string, domainID int, force bool) error {
+
 	if err := execShutdown(getQmpFile(domainName)); err != nil {
-		return logError("failed to execute shutdown command %v", err)
+		return logError("Stop: failed to execute shutdown command %v", err)
+	}
+	//We should quit the domain after shutdown succeeds.
+	//We wait for the domain's state to reach "shutdown" before calling quit.
+	//If domain didn't transition to "shutdown" state within the given time, we stop and quit the domain.
+
+	//waiting for 2 min for state transition
+	delay := time.Second * 120
+	if !waitForStateTransition(domainName, "shutdown", delay) {
```

Comment:
2 min feels arbitrary and I think a better solution here may be to spawn a goroutine that would shepherd the state transitions. That goroutine may just latch onto the qmp control channel in such a way that it exists when the domain gets killed.

That way -- you avoid arbitrary time limit AND unless the domain does the right thing within a global timeout of 10min -- you don't have to worry about killing qemu

Reply:
Changed. 
Spawning a routine everytime we call hyper.Stop(). But couldn't latch onto the socket as the socket supports only one connection at a time, so if I spawn a routine which connects to a socket and waits for the domain state to change, then the method which tries to change the state will not be able to connect as it'll be waiting for the routine to release the connection. 
Instead, the routine checks domain state periodically (every 2 sec). 
If the domain never reaches 'shutdown' state, then the routine exits if the socket is deleted or if it's modified. The latter case is possible when in between the periodic check if we delete and re-create the socket file when the domain is restarting. Also added a timeout of 10 min for the routine.